[Windows] Hide app until first frame is drawn

loic-sharma opened the pull request:
Currently the Windows app is shown immediately with no content and a white background until the first frame is drawn. This change hides the app and makes it visible after that first frame has been drawn. 

Part of https://github.com/flutter/flutter/issues/41980

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

loic-sharma committed:
[Windows] Hide app until first frame is drawn

cbracken reviewed (approved):
![LGTM stamp from a Japanese personal seal](https://user-images.githubusercontent.com/351029/55104787-4e3e7080-5088-11e9-8a12-7cb3789d187e.png)

cbracken commented:
You'll need to request a test exemption from @Hixie.

This code is integration-tested by the devicelab testing. You'll want to sync to tip-of tree and apply this change to the hello_world and platform_view samples that @yaakovschectman recently added Windows devicelab testing for in https://github.com/flutter/flutter/pull/109618.

yaakovschectman reviewed (approved):
LGTM

loic-sharma commented:
I migrated the new Windows app entry points!

Hixie commented:
Can you elaborate on how this is tested?

jonbhanson commented:
@loic-sharma @cbracken in iOS, Android, and Web, my package [flutter_native_splash](https://pub.dev/packages/flutter_native_splash) allows devs to display a splash screen before the first frame has been drawn.  This is preferable to hiding the app because it gives the user immediate feedback that the app is launching and gives a better impression of device responsiveness.

In your development of the Windows Flutter app architecture, can you include the ability for adding a native splash screen that can be displayed before the first frame is drawn?

cbracken commented:
> In your development of the Windows Flutter app architecture, can you include the ability for adding a native splash screen that can be displayed before the first frame is drawn?

In the default app template, we use this to hide the main window for a few additional (and almost certainly unnoticeable) milliseconds until the first frame is ready in order to prevent the user seeing a very short flash of an unrendered view. A splash screen implementor would be able to use this callback to determine _when_ the app is ready to display, which they can use as one input when determining when they'd like to take down the splash screen. 

A splash screen isn't particularly useful for covering up delays measured in milliseconds, hence why it's not default, but:
* The frame callback is something that _is_ potentially useful for splashscreen implementors
* Users of splash screens typically want the main window hidden while the splash screen is up, regardless
* For add-to-app scenarios, or other scenarios where Flutter owns the view, but doesn't own the window, the callback gives the app author a handy way to know when to add the flutter view to their app (or display some other useful signal to the user).

Please shout if any of these patches causes problems for your package.

loic-sharma committed:
Fix

loic-sharma commented:
I rebased this off of https://github.com/flutter/flutter/pull/110114 and added an integration test. @cbracken and @yaakovschectman, I've dismissed your previous approvals given the changes were substantial. Please take another look when available! 😄

yaakovschectman reviewed (approved):
Still looks okay

a-wallen commented on dev/integration_tests/windows_startup_test/lib/main.dart:
```diff
@@ -38,24 +38,32 @@ void main() async {
     final bool? visible = await methodChannel.invokeMethod('isWindowVisible');
-    if (visible == null || visible == false) {
-      throw 'Window should be visible at startup';
+    if (visible == null || visible == true) {
```
Remove `== true` to be concise 
```dart
if( visible == null || visible ) {
  throw 'Window should be hidden at startup';
}
```

If you want to be really concise.
```dart
if( visible ?? true ) {
  throw 'Window should be hidden at startup';
}
```

a-wallen replied:
It's the same for lines 49 and 53 of this file

a-wallen replied:
^^^ nits, don't worry about it.

loic-sharma replied:
I considered both of these before choosing the current approach: using `??` is a little too concise for my taste, and normally I'd drop the `== true` but it looks uneven next to the `== null` check. Let me know if you feel strongly and I'll switch to the `??` approach.

a-wallen replied:
I don't mind as long as it works, which I am sure it does 🙂
